bpo-46337: Urllib.parse scheme-specific behavior without reliance on URL scheme

oldaccountdeadname opened the pull request:
See [bpo-46337](https://bugs.python.org/issue46337). Basically, this allows code like this:

```python
>>> urljoin("custom-scheme://a.host/loc", "..", "custom-scheme://a.host/", classes=[urllib.parse.SchemeClass.RELATIVE, urllib.parse.SchemeClass.NETLOC])
'custom-scheme://a.host/'
```

<!-- issue-number: [bpo-46337](https://bugs.python.org/issue46337) -->
https://bugs.python.org/issue46337
<!-- /issue-number -->

oldaccountdeadname committed:
bpo-46337: Enumerate URL types for urllib's scheme-based behavior

Some features in urllib are dependent on schemes, (i.e., preserving the
netloc in url joining). Prior to this patch, this was governed by the
uses_* lists (uses_relative, uses_netloc, uses_params) which hard code
these attributes for certain schemes. Providing an enum interface and a
'constructor' that allows overrides makes this mechanism a bit more
flexible for future modifications.

oldaccountdeadname committed:
bpo-46337: Allow caller modification of url classes

This allows the callers of urljoin and urlparse to add guaranteed scheme
classes to the url regardless of the actual scheme, which may not be in
the default uses_* lists of schemes. This call-time behavior is done
through an optional parameter that preserves backwards compatibility.

A test case is added for this, and requires the change present in
test_urlparse.checkJoin.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

MaxwellDupre reviewed (changes requested):
Could you add info to docs Please?
Otherwise how will users know how to use or that it exists?

oldaccountdeadname committed:
bpo-46337: Fix grammar of doc comment.

oldaccountdeadname committed:
bpo-46337: document SchemeClass behavior

This functionality was exposed in 53c6ccc.

oldaccountdeadname committed:
add newline to end of news file

It looks like CI expects this when building documentation.

oldaccountdeadname committed:
bpo-46337: fix doctest formatting

oldaccountdeadname committed:
doc: fix some doctests

urljoin will not treat `..` as moving up one directory rather than
moving up one file, thus causing the doctests to fail due to a missing
trailing slash. Both changes are of the form:

http://example.org/post/x -> http://example.org/post/x/

Additionally, the my-protocol example's expected output had the wrong
scheme.

oldaccountdeadname committed:
bpo-43677: fixup PEP 8 style

Minor style things:
+ _scheme_classes' docstring's summary was made explicit.
+ _scheme_classes was prepended with and followed by two newlines.

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -38,13 +39,19 @@
            "parse_qsl", "quote", "quote_plus", "quote_from_bytes",
            "unquote", "unquote_plus", "unquote_to_bytes",
-           "DefragResult", "ParseResult", "SplitResult",
+           "DefragResult", "ParseResult", "SplitResult", "SchemeClass",
```
Following the example of `re.RegexFlag`, name this `SchemeFlag`.  Also, use `enum.Flag` instead of `enum.Enum`.

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -38,13 +39,19 @@
+describe methods for URL resolution, usually by scheme. These resolution classes
+determine, namely, whether a scheme supports, respectively, relative addressing,
+preserving the netloc (domain name), and preserving the parameters."""
+SchemeClass = Enum('SchemeClass', 'RELATIVE NETLOC PARAMS')
```
`SchemeFlag`

JelleZijlstra replied:
Better to use a class statement so the docstring can actually be a docstring.

JelleZijlstra committed:
Merge branch 'main' into urllib-custom-schemes

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -60,6 +67,30 @@
                'mms', 'sftp', 'tel']
 
+
+def _scheme_classes(scheme, overrides=set()):
```
`overrides` should be `options` or `flags`. (`re.search` uses `flags`)

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -363,7 +394,7 @@ def _fix_result_transcoding():
 del _fix_result_transcoding
 
-def urlparse(url, scheme='', allow_fragments=True):
+def urlparse(url, scheme='', allow_fragments=True, classes=set()):
```
`classes=set()` --> `options=SchemeFlag()`

merwok replied:
Did you mean: `options:SchemeFlag=()` ?

ethanfurman replied:
No, `SchemeFlag()` returns an empty flag which can be "updated" by the `|=` further down.  Similarly to:

    var = 0
    var += 7

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -386,7 +417,8 @@ def urlparse(url, scheme='', allow_fragments=True):
     splitresult = urlsplit(url, scheme, allow_fragments)
     scheme, netloc, url, query, fragment = splitresult
-    if scheme in uses_params and ';' in url:
+    scheme_classes = _scheme_classes(scheme, overrides=classes)
```
`_scheme_classes(scheme, overrides=classes)` --> `_scheme_classes(scheme, flags)`

JelleZijlstra commented on Doc/library/urllib.parse.rst:
```diff
@@ -543,6 +555,53 @@ operating on :class:`bytes` or :class:`bytearray` objects:
 
    .. versionadded:: 3.2
 
+Special URL Behaviors and Scheme Classes
```
The name "class" is confusing to me, because it makes me think of a class statement.

Also, why not just add three boolean parameters to urlparse and urljoin? That should make the behavior simpler to use for users.

merwok replied:
See also my alternate proposal for a dedicated function on the ticket.
(Fits with the API design guideline of avoiding a boolean param that changes behaviour — I forget the exact phrasing and origin of that)

ethanfurman replied:
@JelleZijlstra Three boolean parameters would be a pain.  `re.match`, `re.search`, etc., use an `options` parameter to collect all the flags in one.

JelleZijlstra replied:
Why would they be a pain? As a user I'd much rather write `use_relative=True` than look up some separate flags enum.

merwok replied:
Likewise, I am not sure that ~a bitfield~ integer flag is the best thing for a regular Python API.  The re module had a backward compat concern that doesn’t apply here.  But (sorry to repeat my opinion) I am not in favour of piecemeal control, I would rather have a universal parsing function that just implements the universal resource identifier spec :)

ethanfurman replied:
@JelleZijlstra Besides the usual advantages of Enum (repr, verified choices), you also get an easier ability to pass the chosen options around.

ethanfurman replied:
@merwok It's not a bitfield, it's a flag.  ;-)  The three different proposed boolean options all affect the same thing, so a group of choices, a.k.a. an enum, is a good choice -- however, since several choices can be active at once, a flag is a better choice.  Backwards compatibility aside, image if `re.match' looked like:

    def match(need, haystack, ascii=False, multiline=False, ignorecase=False, locale=False, dotall=False, verbose=False, template=False, debug=False):

JelleZijlstra replied:
That sounds pretty nice to me (though the boolean params should be keyword-only). It means I can just look at the signature of `re.match` to know what options I can pass, instead of having to look up a separate set of flags.

It's true though that this sort of API makes it harder to pass around the flags as one unit.

merwok replied:
Same for me!  I don’t remember having to pass re flags around, and if I needed to I could always build a dict for use as varkwargs.

ethanfurman replied:
@merwok (It's not an integer, that would be `IntFlag`)

re the universal parsing function -- how would that differ from the proposed additions to `urlparse.urlparse`?

merwok replied:
See ticket: 

>  urllib is a module that pre-dates the idea of universal parsing for URIs, where the delimiters (like ://) are enough to determine the parts of a URI and give them meaning (host, port, user, path, etc).
> Backward compat for urllib is always a concern; someone said at the time that it could be good to have a new module for modern, generic parsing, but that hasn’t happened.  Maybe a new parse function, or new parameter to the existing one, could be easier to add.

So I don’t want to have to pass parameters to request standard parsing for this or that component, I only want one function that forgets the special registries that urllib.parse relies on and just parses a well-formed URI into the universal components (scheme host path etc).

oldaccountdeadname replied:
> So I don’t want to have to pass parameters to request standard parsing for this or that component, I only want one function that forgets the special registries that urllib.parse relies on and just parses a well-formed URI into the universal components (scheme host path etc).

With the currently pushed code, this would be a one line lambda:

```python
from urllib.parse import SchemeFlag, urlparse

parse_standardized = lambda url: urlparse(url, flags=SchemeFlag.NETLOC | SchemeFlag.RELATIVE | SchemeFlag.PARAMS)
```

Is this worth adding to the PR? I imagine it could be done more extremely (i.e., completely redefine the behavior of urlparse), but would backwards compat not be a concern?

merwok replied:
No, we should not redefine the behaviour of urlparse.

I was always talking about adding another function.  Yes it can be a one-liner, but my point is that I don’t see the usefulness of having the separate flags to pick and choose parts of standard parsing.

ethanfurman replied:
@merwok - added comments to issue; discussion here should be about this PR, discussion about competing PRs (even if the other PRs haven't been written yet ;) should be on the issue tracker.

JelleZijlstra commented on Lib/urllib/parse.py:
```diff
@@ -363,7 +394,7 @@ def _fix_result_transcoding():
 del _fix_result_transcoding
 
-def urlparse(url, scheme='', allow_fragments=True):
+def urlparse(url, scheme='', allow_fragments=True, classes=set()):
```
Avoid mutable defaults, an empty iterable works well. (There are other cases of this throughout the diff.)
```suggestion
def urlparse(url, scheme='', allow_fragments=True, classes=()):
```

ethanfurman replied:
@JelleZijlstra using `SchemeFlag()` creates an immutable default.

JelleZijlstra replied:
I was referring to the original code that used sets, not to your flags suggestion.

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -60,6 +67,30 @@
+    parameter.
+
+    """
+    scheme_classes = set(overrides)
```
remove line

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -60,6 +67,30 @@
+    scheme_classes = set(overrides)
+
+    if scheme in uses_relative:
+        scheme_classes.add(SchemeClass.RELATIVE)
```
`options |= SchemeFlag.RELATIVE`

same transformation for the next two branches

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -60,6 +67,30 @@
+    if scheme in uses_params:
+        scheme_classes.add(SchemeClass.PARAMS)
+
+    return scheme_classes
```
`return options`

ethanfurman reviewed (changes requested):
I think `SchemeFlag` works better than `SchemeClass`.  Either way, use an `enum.Flag` for it, and consider using `__repr__` similar to the one in `re.RegexFlag`.

oldaccountdeadname committed:
urllib: use an enum.FLAG for SchemeClass's.

This commit also contains some miscellaneous doctest fixes.

oldaccountdeadname committed:
urllib: rename SchemeClass to SchemeFlag

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -38,13 +39,25 @@
+    scheme supports, respectively, relative addressing, preserving the
+    netloc (domain name), and preserving the parameters.
+    """
+    NONE = 0
```
We don't need a `NONE` flag.  `SchemeFlag()` or `SchemeFlag(0)` (etc.) will return a zero-valued unnamed flag, which is fine.

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -60,6 +73,27 @@
                'mms', 'sftp', 'tel']
 
+
+def _scheme_classes(scheme, overrides=SchemeFlag.NONE):
```
And this can be `overrides=SchemeFlag()`.

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -348,19 +349,22 @@ or on combining URL components into a URL string.
 
 
-.. function:: urljoin(base, url, allow_fragments=True)
+.. function:: urljoin(base, url, allow_fragments=True, classes=SchemeClass.NONE)
```
There are several places that need `SchemeClass` replaced with `SchemeFlag`.

oldaccountdeadname committed:
urllib: remove SchemeFlag.NONE

SchemeFlag(0) is equivalent.

oldaccountdeadname committed:
doc/urllib: remove old references to SchemeClass

This was renamed to urllib; docs should reflect that.

oldaccountdeadname committed:
news: update news file to use SchemeFlag name

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -543,6 +547,53 @@ operating on :class:`bytes` or :class:`bytearray` objects:
+* ``urllib.uses_params``
+
+In addition, any function that takes a ``classes`` parameter (for instance,
+:func:`urlparse` and :func:`urljoin`) may override the behavior of the uses
```
s/uses/\`\`uses\`\`/

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -543,6 +547,53 @@ operating on :class:`bytes` or :class:`bytearray` objects:
+   'my-protocol://example.org/post/y/'
+
+For reference, the following three scheme classes are present (exactly
+corresponding to the uses lists):
```
s/uses/\`\`uses\`\`/

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -38,13 +39,24 @@
+    RELATIVE = auto()
+    NETLOC = auto()
+    PARAMS = auto()
+
```
change to:

    class SchemeFlag(Flag):
        """SchemeFlag is an enum with the members RELATIVE, NETLOC, and
        PARAMS. These describe methods for URL resolution, usually by
        scheme. These resolution classes determine, namely, whether a
        scheme supports, respectively, relative addressing, preserving the
        netloc (domain name), and preserving the parameters.
        """
        RELATIVE = auto()
        NETLOC = auto()
        PARAMS = auto()
        UNIVERSAL = RELATIVE | NETLOC | PARAMS
        #
        def __repr__(self):
            return f"{self.module}.{self._name_}"
        __str__ = __repr__
    RELATIVE, NETLOC, PARAMS, UNIVERSAL = SchemeFlag

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -511,23 +547,26 @@ def urlunsplit(components):
     """Join a base URL and a possibly relative URL to form an absolute
-    interpretation of the latter."""
+    interpretation of the latter. Some logic may be enabled by setting
+    the classes variable."""
```
s/classes/flags/

gpshead replied:
I also suggest making the new parameter be keyword only.

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -25,7 +25,8 @@ Resource Locators. It supports the following URL schemes: ``file``, ``ftp``,
 ``shttp``, ``sip``, ``sips``, ``snews``, ``svn``, ``svn+ssh``, ``telnet``,
-``wais``, ``ws``, ``wss``.
+``wais``, ``ws``, ``wss``. The behavior of other schemes may be controlled with
+a collection of ``UrlClass`` enums passed to dependent functions.
```
s/UrlClass/SchemeFlag/

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -37,30 +38,35 @@ URL Parsing
       :options: +NORMALIZE_WHITESPACE
 
-      >>> from urllib.parse import urlparse
+      >>> from urllib.parse import urlparse, SchemeFlag
```
s/SchemeFlag/PARAMS/

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -37,30 +38,35 @@ URL Parsing
-      ParseResult(scheme='scheme', netloc='netloc', path='/path;parameters', params='',
-                  query='query', fragment='fragment')
+      ParseResult(scheme='scheme', netloc='netloc', path='/path;parameters', params='', query='query', fragment='fragment')
+      >>> urlparse("scheme://netloc/path;parameters?query#fragment", flags=SchemeFlag.PARAMS)
```
s/SchemeFlag\.//

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -348,19 +349,22 @@ or on combining URL components into a URL string.
 
 
-.. function:: urljoin(base, url, allow_fragments=True)
+.. function:: urljoin(base, url, allow_fragments=True, classes=SchemeFlag.NONE)
```
s/SchemeFlag.NONE/SchemeFlag(0)/

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -543,6 +547,53 @@ operating on :class:`bytes` or :class:`bytearray` objects:
+   >>> from urllib.parse import urljoin, SchemeFlag
+   >>> urljoin(
+   ...     'my-protocol://example.org/post/x/', '../y/',
+   ...     flags=(SchemeFlag.NETLOC | SchemeFlag.RELATIVE))
```
`flags=(NETLOC | RELATIVE)`

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -543,6 +547,53 @@ operating on :class:`bytes` or :class:`bytearray` objects:
+lists, for instance, parsing a custom or widely unused scheme with the same
+behavior as that of HTTP:
+
+   >>> from urllib.parse import urljoin, SchemeFlag
```
s/SchemeFlag/NETLOC, RELATIVE/

ethanfurman commented on Doc/library/urllib.parse.rst:
```diff
@@ -543,6 +547,53 @@ operating on :class:`bytes` or :class:`bytearray` objects:
+
+* ``urllib.SchemeFlag.RELATIVE``
+* ``urllib.SchemeFlag.NETLOC``
+* ``urllib.SchemeFlag.PARAMS``
```
Add `parse` after `urllib` and Remove `SchemeFlag` portion enums will be available at the global level.

Also, add `urllib.parse.UNIVERSAL` and explain (somewhere) what it does.

ethanfurman commented on Lib/test/test_urlparse.py:
```diff
@@ -417,6 +417,11 @@ def test_urljoins(self):
+        self.checkJoin(
+            'nonsensebase://net.loc/url/', '..',
+            'nonsensebase://net.loc/',
+            flags=(urllib.parse.SchemeFlag.RELATIVE | urllib.parse.SchemeFlag.NETLOC),
```
Remove the `SchemeFlag` portion as its enums will be available at the global level.

For those that tend to forget, "global" means "module" in Python.

gpshead replied:
add more test cases that explicitly cover PARAMS and UNIVERSAL behaviors.

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -38,13 +39,24 @@
            "parse_qsl", "quote", "quote_plus", "quote_from_bytes",
            "unquote", "unquote_plus", "unquote_to_bytes",
-           "DefragResult", "ParseResult", "SplitResult",
+           "DefragResult", "ParseResult", "SplitResult", "SchemeFlag",
```
add `"PARAM"`, `"NETLOC"`, `"RELATIVE"`, and `"UNIVERSAL"`

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -60,6 +72,27 @@
+
+    if scheme in uses_params:
+        overrides |= SchemeFlag.PARAMS
+
```
Remove the three `SchemeFlag` references above (again, global level access to the enum members).

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -386,7 +419,8 @@ def urlparse(url, scheme='', allow_fragments=True):
     scheme, netloc, url, query, fragment = splitresult
-    if scheme in uses_params and ';' in url:
+    scheme_classes = _scheme_classes(scheme, overrides=flags)
+    if SchemeFlag.PARAMS & scheme_classes and ';' in url:
```
`if PARAMS in scheme_classes and ';' in url:`

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -500,7 +534,9 @@ def urlunsplit(components):
-    if netloc or (scheme and scheme in uses_netloc and url[:2] != '//'):
+
+    scheme_classes = _scheme_classes(scheme)
+    if netloc or (scheme and (SchemeFlag.NETLOC & scheme_classes) and url[:2] != '//'):
```
if netloc or NETLOC in scheme_classes and url[:2] != '//'):

if `scheme` is empty, then `scheme_classes` will be empty (`SchemeFlag(0)`) so we can just check if `NETLOC` is in `scheme_classes`.

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -511,23 +547,26 @@ def urlunsplit(components):
+    scheme_classes = _scheme_classes(scheme, overrides=flags)
 
-    if scheme != bscheme or scheme not in uses_relative:
+    if scheme != bscheme or not (SchemeFlag.RELATIVE & scheme_classes):
```
if scheme != bscheme or RELATIVE not in scheme_classes:

ethanfurman commented on Lib/urllib/parse.py:
```diff
@@ -511,23 +547,26 @@ def urlunsplit(components):
+    if scheme != bscheme or not (SchemeFlag.RELATIVE & scheme_classes):
         return _coerce_result(url)
-    if scheme in uses_netloc:
+    if SchemeFlag.NETLOC in scheme_classes:
```
if NETLOC in scheme_classes:

ethanfurman reviewed (changes requested):
A few more changes and this should be done!

oldaccountdeadname committed:
doc/urllib: fix formatting and wording

oldaccountdeadname committed:
urllib: expose enums SchemeFlag variants directly

oldaccountdeadname committed:
urllib: add UNIVERSAL SchemeFlag

gpshead commented on Doc/library/urllib.parse.rst:
```diff
@@ -37,30 +38,35 @@ URL Parsing
-      ParseResult(scheme='http', netloc='docs.python.org:80',
-                  path='/3/library/urllib.parse.html', params='',
-                  query='highlight=params', fragment='url-parsing')
+      ParseResult(scheme='http', netloc='docs.python.org:80', path='/3/library/urllib.parse.html', params='', query='highlight=params', fragment='url-parsing')
```
do not reformat doctest examples.  these were formatted to be narrow to avoid horizontal scrollbars in documentation on most common displays and to keep the .rst itself <80 columns when possible.

reformatting is unrelated to the change at hand and distracts from the actual change.

oldaccountdeadname replied:
Restoring the original formatting causes the doctest to fail, I should've broken that out into a separate and clear commit... I'm doctesting with `.python -m doctest`. Is that wrong, or is there some other way I can keep the old linebreaks?

ethanfurman replied:
@gpshead re `behavior_overrides` vs `flags`: aren't flags usually behavior overrides?  `ssl`, `socket`, `_pydecimal`, `_osx_support`, and `re` all use `flags`, while `doctest` uses `compileflags`, `_pyio` use `dec_flags`, and `subprocess` uses `creationflags`.

My first choice here would be a simple `flags`, and it should be easily understood that the flags given will modify the parsing behavior of `urlparse`. Would it be more precise to call it `uri_flags`?  At any rate, `behavior_overrides` is no less generic and much more verbose than `flags`.

gpshead commented on Doc/library/urllib.parse.rst:
```diff
@@ -348,19 +349,22 @@ or on combining URL components into a URL string.
 
 
-.. function:: urljoin(base, url, allow_fragments=True)
+.. function:: urljoin(base, url, allow_fragments=True, classes=SchemeFlag(0))
```
Why is this called `classes` here and `flags` above?  Consistency is important.  These should also be made keyword only arguments.

oldaccountdeadname replied:
I forgot to update after renaming from classes to flags. I'll change it over to `behavior_overrides`.

gpshead commented on Lib/urllib/parse.py:
```diff
@@ -39,12 +40,35 @@
            "parse_qsl", "quote", "quote_plus", "quote_from_bytes",
            "unquote", "unquote_plus", "unquote_to_bytes",
            "DefragResult", "ParseResult", "SplitResult",
+           "SchemeFlag", "RELATIVE", "NETLOC", "PARAMS", "UNIVERSAL",
```
Lets not pollute `__all__` with the CONSTANT_NAMES. People shouldn't really use `from urllib.parse import *` but if they do they shouldn't get these, just SchemeFlag.

oldaccountdeadname replied:
@ethanfurman thoughts? I remember you suggested that these should be exported as such for code like

```python
urlparse(uri_string, flags=UNIVERSAL)
```

or similar. I'm fine either way, but do agree that the namespace would be cleaner were the flags not exported individually.

ethanfurman replied:
Putting them in `globals()` is not for the `from ... import *` case, since, as @gpshead said, folks should not be doing that; putting them in `globals()` is to enable `urlparse.RELATIVE` usage, much like we have `re.IGNORECASE` and not `re.RegexFlag.IGNORECASE`.

gpshead commented on Lib/urllib/parse.py:
```diff
@@ -28,6 +28,7 @@
 """
 
 from collections import namedtuple
+from enum import Flag, auto
```
`import enum` and refer to them as `enum.Flag` and `enum.auto`.

oldaccountdeadname replied:
fixed 0ec4a4edf6

gpshead commented on Lib/test/test_urlparse.py:
```diff
@@ -417,6 +417,11 @@ def test_urljoins(self):
         self.checkJoin('svn+ssh://pathtorepo/dir1', 'dir2', 'svn+ssh://pathtorepo/dir2')
         self.checkJoin('ws://a/b','g','ws://a/g')
         self.checkJoin('wss://a/b','g','wss://a/g')
+        self.checkJoin(
```
make these new SchemeFlag specific test methods instead of appending to an existing long one.

gpshead commented on Lib/urllib/parse.py:
```diff
@@ -363,7 +408,7 @@ def _fix_result_transcoding():
 del _fix_result_transcoding
 
-def urlparse(url, scheme='', allow_fragments=True):
+def urlparse(url, scheme='', allow_fragments=True, flags=SchemeFlag(0)):
```
what does the `0` mean?

ethanfurman replied:
@gpshead Somewhere I thought you said it should be `flags=None` instead -- I agree.

oldaccountdeadname replied:
None is used - `677ed1aac3`.

gpshead reviewed (changes requested):
I agree with Senthil about the API growing something unobvious.  If we add a parameter, it should be keyword only and well named to indicate that it is unusual to provide it.  I'd also set its default value to None rather than introducing the casual user to a new concept (SchemeFlag).

The name `flags=` is probably not sufficient. Something more wordy like `behavior_overrides=` would communicate better that these are not necessary in most cases as reasonable default behaviors exist.  Similarly the name `SchemeFlags` doesn't necessarily communicate the right thing when seen in code as it technically has nothing to do with a scheme itself.  Perhaps something like ParseBehaviors?

oldaccountdeadname committed:
use None rather than SchemeFlag in public API

oldaccountdeadname committed:
do not import from enum

oldaccountdeadname committed:
doc: correct urljoin signature

oldaccountdeadname committed:
make flags parameter keyword-only

oldaccountdeadname committed:
s/classes/flags

merwok commented on Lib/urllib/parse.py:
```diff
@@ -511,23 +562,26 @@ def urlunsplit(components):
     """Join a base URL and a possibly relative URL to form an absolute
-    interpretation of the latter."""
+    interpretation of the latter. Some logic may be enabled by setting
+    the flags variable."""
```
Does this mean: Special behavior can be enabled by passing a flags parameter?

ethanfurman reviewed (changes requested):
See previous comment.
